src/Omega_h_adios2.cpp

```diff
@@ -0,0 +1,566 @@
+  int32_t n;
+  read_value(io, reader, &n, name, true);
+
+  for (int32_t i = 0; i < n; ++i) 
+  {
+    name=pref+"gclas_"+to_string(i)+"_name";
+    adios2::Variable<std::string> bpString = io.InquireVariable<std::string>(name);
+    std::string gclas_name;
+    reader.Get(bpString, gclas_name);
+
+    name=pref+"gclas_"+to_string(i)+"_npairs";
+    int32_t npairs;
+    read_value(io, reader, &npairs, name, true);
+
+    Read<int32_t> gclas_dim = {};
+    Read<int32_t> gclas_id = {};
+
+    name=pref+"gclas_"+to_string(i)+"_dim";
+    read_array(io, reader, gclas_dim, name);
+
+    name=pref+"gclas_"+to_string(i)+"_id";
+    read_array(io, reader, gclas_id,name);
+
+    for (int32_t j = 0; j < npairs; ++j) {
+      ClassPair pair;
+      pair.dim=gclas_dim[j];
+      pair.id=gclas_id[j];
+      mesh->class_sets[gclas_name].push_back(pair);
+    }
+  }
+}
+
+static void write_parents(adios2::IO &io, adios2::Engine &writer, Mesh* mesh, std::string pref)
+{
+  int32_t has_parents = mesh->has_any_parents();
+  std::string name = pref+"has_parents";
+  write_value(io, writer, has_parents, name);
+  if (has_parents) 
+  {
+    for (int32_t d = 0; d <= mesh->dim(); ++d) 
+    {
+      auto parents = mesh->ask_parents(d);
+      name = pref+"parent_"+to_string(d)+"_idx";
+      write_array(io, writer, parents.parent_idx, name);
+      name = pref+"parent_"+to_string(d)+"_codes";
+      write_array(io, writer, parents.codes, name);
+    }
+  }
+}
+
+static void read_parents(adios2::IO &io, adios2::Engine &reader, Mesh* mesh, std::string pref)
+{
+  int32_t has_parents;
+  std::string name = pref+"has_parents";
+  read_value(io, reader, &has_parents, name);
+  if (has_parents) 
+  {
+    for (int32_t d = 0; d <= mesh->dim(); ++d) 
+    {
+      Parents parents;
+      name = pref+"parent_"+to_string(d)+"_idx";
+      read_array(io, reader, parents.parent_idx, name);
+      name = pref+"parent_"+to_string(d)+"_codes";
+      read_array(io, reader, parents.codes, name);
+      mesh->set_parents(d, parents);
+    }
+  }
+}
+
+void write_mesh(adios2::IO &io, adios2::Engine & writer,
+                  Mesh* mesh, std::string pref)
+{
+  comm_size = mesh->comm()->size();
+  rank = mesh->comm()->rank();
+
+  write_meta(io, writer, mesh, pref);
+  int32_t nverts = mesh->nverts();
+  std::string name=pref+"nverts";
+  write_value(io, writer, nverts, name);
+
+  for (int32_t d = 1; d <= mesh->dim(); ++d)
+    write_down(io, writer, mesh, d, pref);
+
+  for (Int d = 0; d <= mesh->dim(); ++d)
+  {
+    write_tags(io, writer, mesh, d, pref);
+    write_part_boundary(io, writer, mesh, d, pref);
+  }
+
+  write_sets(io, writer, mesh, pref);
+  write_parents(io, writer, mesh, pref);
+}
+
+void write(filesystem::path const& path, 
+		  std::map<Mesh*, std::string>& mesh_map)
+{
+  std::map<Mesh*, std::string>::iterator it=mesh_map.begin();
+  Mesh* mesh=it->first;
+
+  if (path.extension().string() != ".bp" && can_print(mesh)) {
+    std::cout
```

Comment:
To avoid spamming stdout, I think this should only be written by one rank.  Does `can_print(...)` return `false` on the non-zero ranks?